### ISSUE TITLE
Reconfiguring file appender not working on Windows

### ISSUE THREAD

**Issue by RichardLewin87** [nomiddlename/log4js-node#473]
Hi,

I am currently trying to reconfigure my file appenders. I am doing this by simply calling the configure function again and passing in the new settings. I have this working as expected on macOS but I am not having the same luck with Windows. On Windows the file appender is not rolling the files correctly after the reconfiguration.

What I am finding is that, when I reconfigure, it will get to a point where the specified maxFileSize gets ignored. Having debugged the code the problem occurs in the streamer library log4js uses when it tries to unlink the first file after reconfiguration that has been rolled over up the max backups specified. No error comes back but the file is not deleted and when right clicking the delete and rename options are shown as being under administrative operations. I believe at this point it has not been fully released by log4js. I believe this is the case as I have observed two things:

1. If I call the shutdown function before reconfiguration I do not have the problem. I therefore assume that the release if files occurs in this function. While this works I then find that the callback from the shutdown function is never called again so I can't use this approach.

2. I am running log4js in an Electron shell. As soon as I close Electron the file is deleted.

All this results in further rollover logic not able to function correctly (I.e. renaming and deleting).

I am using log4js version 1.1.1.

Thank you for any help and guidance on this matter in advance.

**Comment by nomiddlename** [comment]
Yep, when `configure` is called more than once log4js loses its reference to the file handle but doesn't close the file  - which won't then get released until the process ends. Log4js was never really designed to have configure called more than once in an application. That being said, calling `shutdown` before calling configure again is the right way to go - could you provide a small sample of the kind of code you're using? I'm not sure what you mean by the shutdown callback not being called again, so if you could clarify that it would help.

**Comment by RichardLewin87** [comment]
Hi, thanks for your reply. This is how I am using the shutdown function along with the configuration on Windows.

```
log4js.shutdown(function() { 
    setUpLoggers(newSettings);
});
```

setUpLoggers function:

```
log4js.configure({
    appenders: [
        {
            type: 'file',
            layout: {
                    type: "pattern",
                    pattern: "%m"
            },
            filename: clientLogFilePath,
            category: clientLoggerName,
            maxLogSize: clientMaxLogSize,
            backups: clientBackups
        },
        {
            type: 'file',
            layout: {
                    type: "pattern",
                    pattern: "%m"
             },
             filename: appLogFilePath,
             category: appLoggerName,
             maxLogSize: appMaxLogSize,
             backups: appBackups
         },
         {
             type: 'file',
             layout: {
                 type: "pattern",
                  pattern: "[%d] [%p] %m"
             },
             filename: nativeLogFilePath,
             category: nativeLoggerName,
             maxLogSize: nativeMaxLogSize,
             backups: nativeBackups
         }
     ]
});
```

Just to quickly add I had been concentrating on the issue on Windows but have quickly tried the approach of using the shutdown on macOS and it completely breaks the file rolling over process so it's a case of not using the shutdown at all on macOS to ensure all works correctly.

So in the above code it first does go into the callback of shutdown so it does then reconfigure, but the next time shutdown is called the callback is never fired so the reconfiguration never happens.

In digging into the code I can see that in the process of shutting down a list of files to close is created. Then these are 'checked' of as they are successfully closed until all have been closed at which point it calls the callback. However, on the second call to shutdown, I have noticed that the three files from the previous shutdown are still in the list of files to close. Therefore, while it the file count to close is 3 again, the actual list of files is at 6 so it thinks that after it has closed down the 'current' three there are another three to close still.

**Comment by nomiddlename** [comment]
I've changed the way shutdown works in the current version of the code in master, which should help your problem. It's not released to npm yet, but if you're feeling brave you can try it out by changing your package.json dependency to point to the github url (I think, check the npm docs). Be warned that the configuration object structure has changed, but there are docs to help you a little (they're not finished).

**Comment by RichardLewin87** [comment]
Hi,

Thanks for making the change. I have altered my dependency and now on the latest master codebase. Now the shutdown function's callback is being fired more than once and the file rollover is behaving as expected. I have however, come across a new issue.

The main one at the moment is that when I first run the configure function all log log lines after writes to the main 'active' file in what appears to be a locked down state as the file size does not change according to the OS. However, the weird thing is that if I re-run the same configure function it then seems to reset things correctly as the file size changes to reflect what it actually is and a new log line written to the file activates the rollover fully working with the file settings.

I have noticed another oddity but I think it has relation to the above issue.

Do you want me to create a new issue or is it ok to talk further about this new issue in this issue?

**Comment by RichardLewin87** [comment]
Above issue was an error in my code. With regard to the new unreleased version, is there an idea as to when it will be officially released?

**Comment by nomiddlename** [comment]
I want to at least get the documentation up to scratch before I release 2.0 - I know that if I leave the docs for later, I'll never get around to it :-). I'll try to get something out in the next two weeks.

**Comment by nomiddlename** [comment]
Released in version 2.0.0
